### PR TITLE
Leafnode account based connections limits

### PR DESCRIPTION
**Pull request by derekcollison**
Added ability to monitor maximum number of connections for leafnodes and enforce similar to the client connections.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/accounts.go [hunk 1]
```diff
@@ -187,19 +244,25 @@ func (a *Account) TotalSubs() int {
...
 		a.clients[c] = c
 	}
-	if c.kind == SYSTEM {
-		a.sysclients++
+	added := n != len(a.clients)
+	if added {
+		if c.kind == SYSTEM {
+			a.sysclients++
+		}
+		if c.kind == LEAF {
```
Comment (kozlovic) [inline server/accounts.go]: `else` since it can't be both.

File: server/accounts.go [hunk 2]
```diff
@@ -210,11 +273,17 @@ func (a *Account) removeClient(c *client) int {
...
 	n := len(a.clients)
 	delete(a.clients, c)
-	if c.kind == SYSTEM {
-		a.sysclients--
+	removed := n != len(a.clients)
+	if removed {
+		if c.kind == SYSTEM {
+			a.sysclients--
+		}
+		if c.kind == LEAF {
```
Comment (kozlovic) [inline server/accounts.go]: `else` again?

File: server/client.go [hunk 3]
```diff
@@ -444,21 +444,27 @@ func (c *client) registerWithAccount(acc *Account) error {
...
+	c.acc = acc
+	c.applyAccountLimits()
+	c.mu.Unlock()
+
 	// Check if we have a max connections violation
-	if c.kind == CLIENT && acc.MaxTotalConnectionsReached() {
+	if kind == CLIENT && acc.MaxTotalConnectionsReached() {
+		return ErrTooManyAccountConnections
+	}
+	if kind == LEAF && acc.MaxTotalLeafNodesReached() {
```
Comment (kozlovic) [inline server/client.go]: you don't like `else` do you ;-)

Comment (derekcollison) [reply server/client.go]: LOL - I will take a look.

Comment (derekcollison) [reply server/client.go]: I could make them switch statements..

**Comment by kozlovic** [comment]
LGTM

**Pull request merged**